pkg/compiler/symbol_table.go

```diff
@@ -0,0 +1,44 @@
+package compiler
+
+// SymbolScope defines a type of scope that a symbol can be defined inside.
+type SymbolScope string
+
+const (
+	// GlobalScope is the top level scope of an evy program.
+	GlobalScope SymbolScope = "GLOBAL"
+)
+
+// Symbol is a variable inside an evy program.
+type Symbol struct {
+	Name  string
+	Scope SymbolScope
+	Index int
+}
+
+// SymbolTable is a mapping of string identifiers to symbols.
+type SymbolTable struct {
+	store map[string]Symbol
+	// numDefinitions is used to assign each Symbol an increasing index.
+	numDefinitions int
+}
+
+// NewSymbolTable returns a new SymbolTable.
+func NewSymbolTable() *SymbolTable {
+	s := make(map[string]Symbol)
+	return &SymbolTable{store: s}
+}
+
+// Define adds a symbol definition to the table, or increments the index of an already defined symbol.
+func (s *SymbolTable) Define(name string) Symbol {
+	symbol := Symbol{Name: name, Index: s.numDefinitions, Scope: GlobalScope}
+	s.store[name] = symbol
```

Comment:
Without reading the compiler book: it is not clear to us why a symbol gets overriden and how `numDefinitions` is used. will get decremented?

Reply:
I think it's basically just there because you cannot `len` a map, all we want is an increasing index. I was also confused as to why it overwrote symbols with a new index, perhaps just to avoid a lookup? I didn't find any special justification in the book, so we should change the behaviour if it doesn't make sense. wdyt @joshcarp @pgmitche

Reply:
ok, so just a question of - let's role with what the book does for now.
fwiw, you can `len` maps, unless I misunderstand: https://go.dev/play/p/JrKPpLRMUfd ?